[RHDEVDOCS-6821] Pipelines 1.21 release notes (#100818)
openshift/openshift-docs: modules/op-release-notes-1-21-0.adoc

Before this update, the `Pipeline` builder UI failed to save `Buildah` tasks due to a validation error with the default `BUILD_ARGS` parameter. With this update, the validation error is resolved. As a result, the `Pipeline` builder UI correctly saves `Buildah` tasks, even when using the default `BUILD_ARGS` parameter.
+
link:https://issues.redhat.com/browse/SRVKP-8571[SRVKP-8571]

Fixed incorrect sorting of PipelineRuns by duration::
Before this update, sorting `PipelineRuns` by duration used a string sort instead of actual duration values, causing misleading results. With this update, sorting correctly uses the duration in seconds. As a result, `PipelineRuns` are accurately sorted by actual duration.
+
link:https://issues.redhat.com/browse/SRVKP-6211[SRVKP-6211]

Added retry mechanism for TaskRun pod creation on webhook timeout::
Before this update, `TaskRun` pod creation could fail due to a webhook validation timeout, causing `TaskRuns` to be incorrectly marked as failed. With this update, a retry mechanism is added for `TaskRun` pod creation on webhook timeouts. As a result, `TaskRuns` are reliably created and no longer fail due to transient webhook validation issues.
+

> link:https://issues.redhat.com/browse/SRVKP-8835[SRVKP-8835]

Reviewer: Wrong Jira is linked for the retry mechanish for taskrunpod

Reviewer 2: @ochromy is this resolved?

// This module is included in the following assemblies:
// * release_notes/op-release-notes-1-21.adoc

:_mod-docs-content-type: REFERENCE
[id="op-release-notes-1-21-0_{context}"]
= Release notes for {pipelines-title} 1.21

With this update, {pipelines-title} General Availability (GA) 1.21 is available on {OCP} 4.14 and later supported versions.

For more information about the supported versions of {OCP}, see link:https://access.redhat.com/support/policy/updates/openshift#dates[Life Cycle Dates].

[id="new-features-1-21-0_{context}"]
== New features and enhancements

In addition to fixes and stability improvements, the following sections highlight what is new in {pipelines-title} 1.21:

.Pipelines

//https://issues.redhat.com/browse/SRVKP-7303
Read-only root filesystems enabled for {pipelines-shortname} containers::

> With this update, all {pipelines-shortname} containers, including controllers and webhooks, are configured with the `readOnlyRootFilesystem` parameter set to `true`. This change follows security best practices for Kubernetes-based workloads. By enforcing a read-only root filesystem, {pipelines-shortname} improves its security posture by helping to prevent unauthorized modifications to the container runtime environment.

Author: Hello @divyansh42, this is the new single `readOnlyRootFilesystem` RN.  PTAL. 
Thank you!

.Operator

//Example missing - https://issues.redhat.com/browse/SRVKP-6166
`legacyPipelineRbac` parameter for controlling pipeline service account permissions::
* Before this update, the pipeline service account automatically received the `edit` ClusterRole within its namespace, following the legacy RBAC behavior.
With this update, you can use the new `legacyPipelineRbac` parameter to control permissions:

** Set to `true` (default) to preserve the legacy behavior.

** Set to `false` to prevent the pipeline service account from receiving the `edit` ClusterRole, enforcing more restricted permissions by default.

+
[IMPORTANT]
====
Existing role bindings are not automatically removed from namespaces with the pipeline service account. You must remove them manually when changing this parameter on existing deployments.
====

.{pac}

//https://issues.redhat.com/browse/SRVKP-9056
Improved performance for GitLab project access control checks::

> * With this update, {pac} caches the results of GitLab Project Access Control List (ACL) membership checks.

Reviewer 3: ```suggestion
* With this update, {pac} caches the results of GitLab Project Access Control List (ACL) membership results.
```

> This optimization reduces repeated API calls to the GitLab service, improving performance and efficiency during permission validation.

Reviewer 3: ```suggestion
This optimization reduces repeated API calls to the GitLab service, improving performance and efficiency during permission check.
```

                  matchStatuses:
                    - "Failed"
                retention: "180d"
              - name: "retain-annotated-for-debug"
                selector:
                  matchAnnotations:
                    "debug/retain": ["true"]
                retention: "14d"
              - name: "default-production-policy"
                selector:
                  matchNamespaces:
                    - "production"
                    - "prod-east"
                retention: "60d"
              - name: "short-term-ci-retention"
                selector:
                  matchNamespaces:
                    - "ci"
                retention: "10h"
----
+
This config map example:

** Runs the pruning job daily at 2:00 AM, as specified by the `runAt` cron schedule.

** Keeps failed Results in `production` or `prod-east` with `criticality: high` for 180 days.

** Keeps Results with annotation `debug/retain: "true"` for 14 days.

** Keeps other Results in `production` or `prod-east` for 60 days.

** Keeps Results in the `ci` namespace for 10 hours.

** Keeps all other Results for the default retention period (30 days).

.{tekton-cache}

> 

Reviewer 4: Tekton-cache is going GA with this release. So please mention that as well

Author: Included.
Thanks!

//https://issues.redhat.com/browse/RHDEVDOCS-6985 
{tekton-cache} is generally available::
With this update, {tekton-cache} is generally available (GA) and is fully supported for production use. {tekton-cache} was previously available as a Technology Preview (TP) feature.

//https://issues.redhat.com/browse/SRVKP-9151
Tekton Cache binaries available for public download::
With this update, the {tekton-cache} product binaries are available for download without authentication. This accessibility enables customers to use the Red Hat binaries for their custom `StepAction` configurations.

//Possibly missing example? - https://issues.redhat.com/browse/SRVKP-8303
Improved support for Docker credentials::
Before this update, `{tekton-cache}` required Docker secrets to include a `config.json` key. With this update, Docker secrets without a `config.json` key are supported. The `DOCKER_CONFIG` parameter can point to any location containing either a `config.json` file or a `.dockerconfigjson` file, improving flexibility for private registry authentication.

.{manual-approval-gate}

//Possibly missing example? - https://issues.redhat.com/browse/SRVKP-7187 
Group support for Approval Task::

> With this update, the Approval Task supports group approvers. Specify group approvers using the `group:<groupName>` syntax in the `params` list. Any member of the group can approve or reject the task. An approval by any group member counts as a single approval, while a rejection by any member immediately counts as a single rejection and fails the task. This enhancement provides more flexible approval workflows by allowing teams to delegate approvals to groups instead of individuals.

Reviewer 5: ```suggestion
With this update, the Approval Task supports group approvers. Specify group approvers using the `group:<groupName>` syntax in the `params` list. Any member of the group can approve or reject the task. An approval by any group member counts as a single approval, while a rejection by any member immediately counts as a single rejection and fails the task. This enhancement provides more flexible approval workflows by allowing teams to delegate approvals to groups as well as individuals.
```

+
[source, yaml]
----
apiVersion: tekton.dev/v1
  kind: PipelineRun
  metadata:
    name: example-with-timeout-override
  spec:
    pipelineRef:
      name: my-pipeline
    timeouts:
      pipeline: "2h"
    taskRunSpecs:
      - pipelineTaskName: build-task
        timeout: "30m"
      - pipelineTaskName: test-task
        timeout: "1h"
----
+
This allows finer-grained control over task execution duration without affecting the overall `PipelineRun` timeout.

//https://issues.redhat.com/browse/SRVKP-8880
//Unknown component
Resolver caching for bundle, Git, and cluster resolvers::
+
With this update, resolver caching is supported for bundle, Git, and cluster resolvers. This helps reduce redundant fetches, minimize external API calls, and improve pipeline execution reliability, especially when external services impose rate limits or are temporarily unavailable.
+
* You can configure caching using the `resolver-cache-config` config map, where you can set the cache size and adjust the time to live (TTL) value without restarting controllers:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: resolver-cache-config
data:
  max-size: "1000"
  ttl: "5m"
#...
----
+

> ** `max-size`: the maximum amount of cached entries. The default value is `1000`.

Author: Good catch, thank you @aThorp96!

> ** `ttl`: the time to live of the cache entry. The default value is `5m`.

Reviewer 2: @ochromy please resolve if incorporated.

//

.{pac}

//https://issues.redhat.com/browse/SRVKP-9056
Improved performance for GitLab project access control checks::
With this update, `{pac}` caches the results of GitLab Project Access Control List (ACL) membership queries. This optimization reduces repeated API calls to the GitLab service, improving performance and efficiency during permission checks.

//Example Missing - https://issues.redhat.com/browse/SRVKP-8900 and SRVKP-8165
Configure the number of lines in error log snippets::
With this update, you can configure how many lines appear in error log snippets using the new `error-log-snippet-number-of-lines` setting:
+
[source, yaml]
----
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonConfig
metadata:
  name: config
spec:
  platforms:
    openshift:
      pipelinesAsCode:
        enable: true
        settings:
          error-log-snippet-number-of-lines: '3'
          #...
----
+
Log snippets are automatically truncated to 65,000 characters, preventing failures when posting check-run updates to the GitHub API.

//https://issues.redhat.com/browse/SRVKP-9315
SHA validation added to `/ok-to-test` commands to prevent race conditions::

> With this update, {pac} implements the `require-ok-to-test-sha` feature, which enforces commit SHA validation for `/ok-to-test` commands. This helps mitigate potential Time-of-Check to Time-of-Use (TOCTOU) race conditions in Continuous Integration (CI) workflows.

Reviewer 3: ```suggestion
With this update, {pac} implements the `require-ok-to-test-sha` feature, which enforces commit SHA value with  `/ok-to-test` commands. This helps mitigate potential Time-of-Check to Time-of-Use (TOCTOU) race conditions in Continuous Integration (CI) workflows.
```

                    - "ci"
                retention: "10h"
----
+
This config map example:

* Runs the pruning job daily at 2:00 AM, as specified by the `runAt` cron schedule.
* Keeps failed Results in `production` or `prod-east` with `criticality: high` for 180 days.
* Keeps Results with annotation `debug/retain: "true"` for 14 days.
* Keeps other Results in `production` or `prod-east` for 60 days.
* Keeps Results in the `ci` namespace for 10 hours.
* Keeps all other Results for the default retention period (30 days).

//https://issues.redhat.com/browse/SRVKP-8057
Automatic route creation for the Results API endpoint::
With this update, {tekton-results} automatically creates a route for the Results API endpoint. This improves accessibility by exposing the API externally without requiring additional user configuration. You can optionally provide a custom host and path to tailor the route to your environment.

//https://issues.redhat.com/browse/SRVKP-8006
PostgreSQL support updated to version 17.5::
With this update, {tekton-results} supports PostgreSQL version 17.5.

//https://issues.redhat.com/browse/SRVKP-8157
New metrics for runs not stored in the database::

> With this update, `{tekton-results}` adds the `runs_not_stored_count` metric. This metric tracks the number of `PipelineRun` and `TaskRun` instances that are deleted before they can be persisted in the database. Supported tags include:

Author: Changed to: 

> With this update, `{tekton-results}` adds the `runs_not_stored_count` metric, emitted by the default `watcher` container. This metric ...

+
link:https://issues.redhat.com/browse/SRVKP-8377[SRVKP-8377]

Corrected `prioritySemaphore` locking to prevent deadlocks and race conditions::
Before this update, the `prioritySemaphore` implementation could cause deadlocks, race conditions, and panics due to unsynchronized data access. With this update, the locking logic is corrected and all shared data is properly synchronized, preventing these concurrency issues.
+
link:https://issues.redhat.com/browse/SRVKP-8198[SRVKP-8198]

Retained pods on `TaskRun` timeout when `keep-pod-on-cancel` is true::
Before this update, pods were retained only if a `TaskRun` was canceled. With this update, pods are also retained when a `TaskRun` times out if `keep-pod-on-cancel` is set to `true`.
+
link:https://issues.redhat.com/browse/SRVKP-9176[SRVKP-9176]

Fixed immediate YAML editor updates using `useEffect` hook::
Before this update, YAML editor changes were not applied until the component was remounted. With this update, changes are immediately reflected using the `useEffect` hook, improving the editing experience.
+

> link:https://issues.redhat.com/browse/SRVKP-8205[SRVKP-8205]

Author: Updated.

Before this update, TektonConfig container args could contain duplicates and `["-key","value"]` pairs. With this update, flags are normalized to the `"-key=value"` format and duplicates are removed, simplifying configuration.
+
link:https://issues.redhat.com/browse/SRVKP-7889[SRVKP-7889]

`PipelineRuns` fail clearly on invalid `taskRef.apiVersion`::
Before this update, setting the `*.spec.tasks[*].taskRef.apiVersion` field to an invalid value caused `PipelineRun` execution to fail silently. With this update, `PipelineRuns` throw a clear error when `taskRef.apiVersion` is invalid.
+
link:https://issues.redhat.com/browse/SRVKP-8514[SRVKP-8514]

Git resolver correctly trusts OpenShift Proxy's custom PKI::
Before this change, a regression caused the Git resolver to no longer use the OpenShift Proxy's custom-configured public key infrastructures (PKIs). This issue could prevent the Git resolver from resolving references to a self-hosted Git provider. With this update, the OpenShift Proxy's full CA bundle is trusted by all components, including the Git resolver, which trusts any certificates configured in the cluster's custom PKI.
+

> link:https://issues.redhat.com/browse/SRVKP-9177[SRVKP-9177]

Reviewer 2: Yeah, we don't need to include in 1.21
cc @ochromy

Support for rerunning resolver-based PipelineRuns::
With this update, the `tkn` CLI introduces a new `--resolvertype` flag to the `tkn p start` command. This flag allows you to specify the resolver type, such as `git`, `http`, `hub`, `cluster`, `bundle`, or `remote`, for any supported resolver type, when re-running a resolver-based `PipelineRun`. You can reference an existing `PipelineRun` name to re-run it using the specified resolver type.

//https://issues.redhat.com/browse/SRVKP-9142
New --resolvertype to support rerunning resolver-based PipelineRuns::
With this update, the `tkn p start --last` command introduces the `--resolvertype` flag. This flag enables users to specify the resolver type, such as `git`, `hub`, or `bundle`, when re-running a previous resolver-based `PipelineRun`. Additionally, the help text for the command has been updated to use the correct pronoun.

.Pruner

//https://issues.redhat.com/browse/SRVKP-6989
Event-driven pruner is generally available::
With this update, the event-driven pruner `tektonpruner` is generally available (GA) and is fully supported as a pruning mechanism for {pipelines-shortname} with centralized and hierarchical configuration.
+

> While existing pruning mechanisms, such as the default job-based pruner, {pac} `keep-max-run`, and {tekton-results} based `retention`, continue to function, uers currently relying on legacy pruning approaches are encouraged to adopt the event-driven pruner `tektonpruner` to help ensure smoother performance, more predictable cleanup, and reduced operational overhead.

Reviewer: ```suggestion
While existing pruning mechanisms, such as the default job-based pruner, {pac} `keep-max-run`, and {tekton-results} based `retention`, continue to function, users currently relying on legacy pruning approaches are encouraged to adopt the event-driven pruner `tektonpruner` to help ensure smoother performance, more predictable cleanup, and reduced operational overhead.
```

// This module is included in the following assemblies:
// * release_notes/op-release-notes-1-21.adoc

:_mod-docs-content-type: REFERENCE
[id="op-release-notes-1-21-0_{context}"]
= Release notes for {pipelines-title} 1.21

> With this update, {pipelines-title} General Availability (GA) 1.21 is available on {OCP} 4.15 and later versions.

Reviewer: ```suggestion
With this update, {pipelines-title} General Availability (GA) 1.21 is available on {OCP} 4.16 and later versions.
```

Reviewer 2: It is " 4.14 and then 4.16 onwards"

Reviewer: 4.14 and 4.16 onwards

.Pipelines

//https://issues.redhat.com/browse/SRVKP-8341
Override individual TaskRun timeouts in a PipelineRun::
With this update, you can override the timeout for individual tasks within a `PipelineRun` using the `spec.taskRunSpecs[].timeout` field.
+
[source, yaml]
----
apiVersion: tekton.dev/v1
  kind: PipelineRun
  metadata:
    name: example-with-timeout-override
  spec:
    pipelineRef:
      name: my-pipeline
    timeouts:
      pipeline: "2h"
    taskRunSpecs:
      - pipelineTaskName: build-task
        timeout: "30m"
      - pipelineTaskName: test-task
        timeout: "1h"
----
+
This allows finer-grained control over task execution duration without affecting the overall `PipelineRun` timeout.

//https://issues.redhat.com/browse/SRVKP-8880
Resolver caching for bundle, Git, and cluster resolvers::
+
With this update, resolver caching is supported for bundle, Git, and cluster resolvers. This helps reduce redundant fetches, minimize external API calls, and improve pipeline execution reliability, especially when external services impose rate limits or are temporarily unavailable.
+

> * You can configure caching using the `resolver-cache-config` config map, where you can set the cache size and adjust the time to live (TTL) value without restarting controllers:

Reviewer 2: Customizing the configuration to update the default values should happen through the `options` section of `tektonconfig` for pipelines spec.
The issue of documenting in downstream docs to guide users directly update the configmap will result in issues during upgrade.
@waveywaves could you help to update the sample using tektonconfig. 
We can probably refer sample
```
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonConfig
metadata:
  name: config
spec:
  pipeline:
    options:
      configMaps:
        resolver-cache-config:
          data:
            max-size: "1000"
            ttl: "5m"
```

Reviewer 2: @waveywaves can you please take a look at this?

+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: resolver-cache-config
data:
  max-size: "1000"
  ttl: "5m"
#...
----
+
** `max-size`: the maximum number of cached entries. The default value is `"1000"`.
** `ttl`: the time to live of the cache entry. The default value is `"5m"`.
+

> * Caching mode can be set globally in the resolver config maps, such as `bundleresolver-config`, `git-resolver-config`, `cluster-resolver-config`, or on the task-level:

Reviewer 2: Same as the above comment, we should not edit the configmap

+
[source,yaml]
----
#...
data:
  cache: "auto"
  #...
----
+
** The available modes are `auto` (cache only immutable references), `always` (cache everything), and `never` (disable caching). The default value is `auto`.
+
* Task-level override:
+
[source,yaml]
----
#...
params:
  - name: cache
    value: "always"
#...
----

+
Resolver caching helps improve reliability, reduce latency for frequently accessed resources, and decrease load on external services such as GitHub and OCI registries. Cache hits, misses, and timestamps are added to resource annotations for resolved resources.

> Array values can be resolved in when expressions::

Reviewer 2: @waveywaves can you please add an example since this is a new feature. 
@srivickynesh FYI this is an upstream feature, if possible can we do a basic test as it is customer facing

Reviewer 2: @ochromy please resolve if incorporated.

With this update, array value resolution is enabled in the `Input` attribute of `When` expressions.
//+
//link:https://issues.redhat.com/browse/SRVKP-9135[SRVKP-9135]

> Support for display names added to steps::

Reviewer 2: @waveywaves can you please provide a sample for this as well.

Reviewer 2: @waveywaves can you please take a look at this?

Reviewer 6: ```
apiVersion: tekton.dev/v1                                                                               
kind: Task                                                                                              
metadata:                                                                                               
  name: build-and-test                                                                                  
spec:                                                                                                   
  params:                                                                                               
    - name: app-name                                                                                    
      type: string                                                                                      
  steps:                                                                                                
    - name: build                                                                                       
      displayName: "Build the application"                                                              
      image: golang:1.21                                                                                
      script: |                                                                                         
        go build ./...                                                                                  

    - name: test                                                                                        
      displayName: "Run unit tests for $(params.app-name)"  # Supports param substitution                                                                                                                          
      image: golang:1.21                                                                                
      script: |                                                                                         
        go test ./...                                                                                   

    - name: lint                                                                                        
      displayName: "Lint source code"                                                                   
      image: golangci/golangci-lint                                                                     
      script: |                                                                                         
        golangci-lint run
```

          on_cel: 'body.pipelineRun.status.conditions[0].reason == "Failed"'
          context_items:
            error_content: true
            container_logs:
              enabled: true
              max_lines: 100
          output: "pr-comment"
  #...
----
* `model`: Optional. If you do not provide a value, the framework uses the provider default.

//link to downstream docs

//https://issues.redhat.com/browse/SRVKP-8363 and https://issues.redhat.com/browse/SRVKP-8612
Trigger and cancel PipelineRuns for Git tags::
With this update, you can trigger `PipelineRuns` by using comments on commits associated with a Git tag. This provides flexible, version-specific control of CI/CD workflows.
+
To trigger a `PipelineRun`, add a comment on the tagged commit using the format `/test <pipeline-name> tag:<tag-name>`, for example:
+
[source]
----
/test xyz-pipeline-run tag:v1.0.0
----
+
To cancel a `PipelineRun`, use the format `/cancel <pipeline-name> tag:<tag-name>`, for example:
+
[source]
----
/cancel xyz-pipeline-run tag:v1.0.0
----
+
This feature is supported for GitHub and GitLab, enabling teams to trigger or cancel `PipelineRuns` tied to specific tagged versions.

> //https://issues.redhat.com/browse/SRVKP-8313

Reviewer 2: @savitaashture can you please confirm here

> //https://issues.redhat.com/browse/SRVKP-8313

Reviewer 2: @khrm can you please provide your opinion here?

ManagedBy support for PipelineRuns and TaskRuns::
With this update, the `PipelineRun` and `TaskRun` resources include a `spec.managedBy` configuration field. You can use this field to delegate responsibility for managing the lifecycle of the run to an external controller or operator.
+
* The field has no default value and must be explicitly set to take effect.
* If the field is set and does not reference the built-in controller, reconciliation by the default controller is skipped.
* The field is immutable and cannot be changed after the resource is created.

+
This feature enables external systems to control specific `PipelineRun` and `TaskRun` resources without interference from the built-in controller.

New security policy mitigates TOCTOU vulnerability::

> A new policy, `require-ok-to-test-sha`, is introduced in {pac} to enforce commit SHA validation when using the `/ok-to-test` comment command on GitHub pull requests. This feature mitigates a critical TOCTOU (Time-of-Check to Time-of-Use) race condition vulnerability, specific to the GitHub provider, where an attacker could execute a pipeline on an unapproved SHA. When the policy is enabled, users must specify the exact commit SHA, for example, `/ok-to-test <sha>`, for approval. As a result, this ties the approval directly to a specific commit, preventing pipeline execution on any subsequent malicious force-pushed SHA.

Reviewer: Duplicate section of line 188

+
* `kind` – the type of run (`PipelineRun` or `TaskRun`),
* `namespace` – the namespace where the run was created.

+
For example:
[source, terminal]
----
watcher_runs_not_stored_count{kind="PipelineRun",namespace="default"} 5
----

//https://issues.redhat.com/browse/SRVKP-8078
Metrics for run storage latency::
With this update, {tekton-results} adds the `run_storage_latency_seconds` metric, emitted by the default `watcher` container. This metric measures the time between run completion and its successful storage in the database.
+
Supported tags include:
+
* `kind` – the type of run (`PipelineRun` or `TaskRun`)
* `namespace` – the namespace where the run was created
+
[source]
----
watcher_run_storage_latency_seconds{kind="PipelineRun",namespace="default"} 0.5
----
+
This metric is emitted only when a run transitions from completed to stored, helping ensure accurate measurement of storage latency without being skewed by multiple reconciliations.

CLI configuration persists across namespaces::

> Before this update, the API configuration for the {tekton-results} CLI was fetched from the current namespace context. As a consequence, users had to re-authenticate or reconfigure after switching namespaces. With this update, the {oc-first} {tekton-results} configuration is fetched from the user context field instead of the namespace context. As a result, the configuration persists across namespace switches, removing the need to run the `opc results config set` command repeatedly.

Reviewer 2: It should be `opc` not `oc`

Author: Changed to:
```suggestion
Before this update, the API configuration for the {tekton-results} CLI was fetched from the current namespace context. As a consequence, users had to re-authenticate or reconfigure after switching namespaces. With this update, the configuration persists across namespace switches, removing the need to run the `opc results config set` command repeatedly.
```

Default database migration to PostgreSQL version 15::
The PostgreSQL image used by the default {tekton-results} deployment is upgraded from version 13 to version 15, addressing the upcoming end of life (EOL) for PostgreSQL 13. This process implements an automated migration, which uses the slower yet more reliable data copy mechanism from the original PostgreSQL image.
+
[IMPORTANT]
====
If you are using the default PostgreSQL deployment, ensure you have backed up your data and that the underlying Persistent Volume Claim (PVC) has more than 50% free space before you start the {pipelines-shortname} upgrade. If you are using an external database with {tekton-results}, you are not affected by this change.
====
//+
//link:https://issues.redhat.com/browse/SRVKP-9336[SRVKP-9336]

.{tekton-cache}

//https://issues.redhat.com/browse/RHDEVDOCS-6985
{tekton-cache} is generally available::
With this update, {tekton-cache} is generally available (GA) and is fully supported for production use. {tekton-cache} was previously available as a Technology Preview (TP) feature.

//https://issues.redhat.com/browse/SRVKP-9151
Tekton Cache binaries available for public download::

> With this update, the {tekton-cache} product binaries are available for download without authentication. This accessibility enables customers to use the Red Hat binaries for their custom `StepAction` configurations.

Reviewer 2: @pramodbindal can you please take a look?

//REFERENCE DOC LINK

//https://issues.redhat.com/browse/SRVKP-8991 and SRVKP-8069
Namespace-level pruner configuration::
With this update, the event-driven pruner supports custom pruning policies at the namespace level. You can define custom pruning policies, such as time to live and history limits, that override global defaults by creating a `tekton-pruner-namespace-spec` config map in your namespace.
//REFERENCE DOWSTREAM DOC LINK

//https://issues.redhat.com/browse/SRVKP-7744
Selector-based pruning configuration::
With this update, the event-driven pruner supports selector-based resource matching with the `matchLabels` and `matchAnnotations` selectors. The selector matching logic when you specify both `matchLabels` and `matchAnnotations` selectors is AND and name matching has absolute precedence regardless of selector presence. The selector-based resource matching is supported only in the namespace-level `tekton-pruner-namespace-spec` config maps, not in the global `tekton-pruner-default-spec` config maps.
//REFERENCE DOWSTREAM DOC LINK

//https://issues.redhat.com/browse/SRVKP-9253

> System-wide maximum limits added for the event-driven pruner configuration::

Reviewer 2: @anithapriyanatarajan any update on this?

+
link:https://issues.redhat.com/browse/SRVKP-8922[SRVKP-8922]

Webhook validation no longer targets control-plane namespaces::
Before this update, the logic for the `tekton-operator-proxy-webhook` parameter attempted to validate resources in control-plane namespaces, such as `kube-\*` and `openshift-\*`. This behavior caused unintended webhook certificate issues that affected unrelated system components. With this update, the webhook logic excludes all control-plane namespaces from admission validation. This improvement ensures better isolation between Tekton components and other cluster operators.
+
link:https://issues.redhat.com/browse/SRVKP-8891[SRVKP-8891]

Custom hub catalog configuration is preserved during conversion::
Before this update, the {pipelines-shortname} Operator removed the `catalog-{INDEX}-type` field during conversion, which caused the loss of custom hub catalog types. With this update, the Operator preserves the `catalog-{INDEX}-type` field in its config map.
+
link:https://issues.redhat.com/browse/SRVKP-9472[SRVKP-9472]

> 

Reviewer 7: We need to include a fixed issue related to Pruner related to - https://issues.redhat.com/browse/SRVKP-8621. This was mentioned as a known issue in 1.20.x

Reviewer 2: @anithapriyanatarajan can you please take a look?

//https://issues.redhat.com/browse/SRVKP-8906
Group support for Approval Tasks::
With this update, Approval Tasks support group approvers. You can specify a group using the `group:<groupName>` syntax in the `params` list. Any member of the group can approve or reject the task. Approvals by any group member count as a single approval, while a rejection by any member immediately counts as a single rejection and fails the task. Group members also receive notifications about the task, just like individual approvers.

Pipeline Overview page retains filter selections::
The Pipeline Overview page persists user selections for the *Namespace*, *Time Range*, and *Refresh Interval* filters. These selections are stored in the application state and URL query parameters. As a result, users have a consistent experience when navigating away, returning to the page, or refreshing the browser. The filters reset only when the user switches namespaces.
//+
//link:https://issues.redhat.com/browse/SRVKP-9607[SRVKP-9607]

Time-range filter label updated for clarity::

> The time-range filter previously labeled "Last weeks" is updated to "Last week". This change resolves customer confusion regarding the intended single-week time range and improves usability. As a result, the corresponding API payload is aligned to accurately reflect the 7-day range, ensuring consistency between the UI and the {tekton-results} API.

Reviewer 2: > As a result, the corresponding API payload is aligned to accurately reflect the 7-day range, ensuring consistency between the UI and the Tekton Results API.

@anwesha-palit-redhat do you think for a user this technical clarification is needed?

Reviewer 8: @divyansh42 Yes the purpose for this is that some customers might be used to using the previous filter which was set to 2 weeks instead of 1 week. So if we do not explain this change they might raise a question regarding why this change was introduced.

//+
//link:https://issues.redhat.com/browse/SRVKP-9428[SRVKP-9428]

.{pac}

//https://issues.redhat.com/browse/SRVKP-9056
Improved performance for GitLab project access control checks::
With this update, `{pac}` caches the results of GitLab Project Access Control List (ACL) membership queries. This optimization reduces repeated API calls to the GitLab service, improving performance and efficiency during permission checks.

//https://issues.redhat.com/browse/SRVKP-8900 and SRVKP-8165
Configure the number of lines in error log snippets::
With this update, you can configure how many lines appear in error log snippets using the new `error-log-snippet-number-of-lines` setting:
+
[source, yaml]
----
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonConfig
metadata:
  name: config
spec:
  platforms:
    openshift:
      pipelinesAsCode:
        enable: true
        settings:
          error-log-snippet-number-of-lines: "3"
          #...
----
+

> Log snippets are automatically truncated to 65,000 characters, preventing failures when posting check-run updates to the GitHub API.

Reviewer 3: it is only for GitHub

----
* `-b` or `--body`: Specify a path to a JSON body file. This is a webhook payload.
* `-H` or `--headers`: Specify a path to headers file. This can be a plain text, a JSON file, or a gosmee script.
* `-p` or `--provider`: Specify the provider. This can be `github`, `gitlab`,  `bitbucket-cloud`, `bitbucket-datacenter`, `gitea`, or `auto` for automatic detection from the payload.

+
Key capabilities include:
* Interactive mode: Provides a prompt in the terminal to type CEL expressions, with tab completion for variables and payload fields.
* Variable access using:
** direct variables, such as `event` or `target_branch`,
** webhook payload fields, for example, `body.action`,
** HTTP headers, for example, `headers['X-GitHub-Event']`,
** {pac} parameters, for example, `pac.revision`.
* Debugging: Quickly test and debug CEL expressions used in `PipelineRun` resource configurations against real webhook data.
+
link:https://issues.redhat.com/browse/SRVKP-8026[SRVKP-8026]

[id="pipelines-breaking-changes-1-21-0"]
== Breaking changes

.{pac}

//https://issues.redhat.com/browse/SRVKP-8663
pipelinerun_status field in Repository custom resource is depracated::

> With this update, the `pipelinerun_status` field of the `Repository` custom resource (CR) is deprecated. Update any integrations or automation that reference this field to ensure compatibility with upcoming versions.

Reviewer: @zakisk I see differently in releasenotes in Jira, can you confirm?

Configure the number of lines in error log snippets::
With this update, you can configure how many lines appear in error log snippets using the new `error-log-snippet-number-of-lines` setting:
+
[source, yaml]
----
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonConfig
metadata:
  name: config
spec:
  platforms:
    openshift:
      pipelinesAsCode:
        enable: true
        settings:
          error-log-snippet-number-of-lines: "3"
          #...
----
+
Log snippets are automatically truncated to 65,000 characters, preventing failures when posting check-run updates to the GitHub API.

//https://issues.redhat.com/browse/SRVKP-8908
GitLab commit status set on the source and target project by default::
With this update, {pac} attempts to set the commit status on both the source and target `upstream` projects in GitLab. If both attempts fail due to permissions issues, the system falls back to posting a comment. This helps ensure that the commit status is reported in the most relevant location for merge requests.

//https://issues.redhat.com/browse/SRVKP-8902
Improved error message for GitLab private repository access failures::

> With this update, {pac} proactively checks whether the configured GitLab token has the required read access to the source repository when processing merge requests from private forks. If the token lacks the necessary `read_repository` scope, {pac} fails early with a clear, actionable error message. This helps ensure that permission issues are easier to identify and prevents pipelines from failing unexpectedly later in the process.

Reviewer 3: or what do you mean by that? I think I didn't understand it fully

> With this update, {pac} proactively checks whether the configured GitLab token has the required read access to the source repository when processing merge requests from private forks. If the token lacks the necessary `read_repository` scope, {pac} fails early with a clear, actionable error message. This helps ensure that permission issues are easier to identify and prevents pipelines from failing unexpectedly later in the process.

Reviewer 2: I think the above error message is fine. @ochromy is it possible to reframe the RN text with the above error message?

//https://issues.redhat.com/browse/SRVKP-8577

> AI integration for automated pull request analysis::

Author: Looking at it now, we already have the Lightspeed integration through `opc assist` documented downstream for this release. 
Do we even want to keep this in, since this is a "foundation" for integration with Lightspeed?

> AI integration for automated pull request analysis::

Author: Added to PaC Tech Preview section.

+
[source]
----
/test xyz-pipeline-run tag:v1.0.0
----
+
To cancel a `PipelineRun`, use the format `/cancel <pipeline-name> tag:<tag-name>`, for example:
+
[source]
----
/cancel xyz-pipeline-run tag:v1.0.0
----
+
This feature is supported for GitHub and GitLab, enabling teams to trigger or cancel `PipelineRuns` tied to specific tagged versions.

//https://issues.redhat.com/browse/SRVKP-8313
ManagedBy support for PipelineRuns and TaskRuns::
With this update, the `PipelineRun` and `TaskRun` resources include a `spec.managedBy` configuration field. You can use this field to delegate responsibility for managing the lifecycle of the run to an external controller or operator.
+
* The field has no default value and must be explicitly set to take effect.
* If the field is set and does not reference the built-in controller, reconciliation by the default controller is skipped.
* The field is immutable and cannot be changed after the resource is created.

+
This feature enables external systems to control specific `PipelineRun` and `TaskRun` resources without interference from the built-in controller.

SHA validation added to /ok-to-test commands to prevent race conditions::

> A new policy, `require-ok-to-test-sha`, is introduced in {pac} to enforce commit SHA validation when using the `/ok-to-test` comment command on GitHub pull requests. This feature mitigates a critical TOCTOU (Time-of-Check to Time-of-Use) race condition vulnerability, specific to the GitHub provider, where an attacker could execute a pipeline on an unapproved SHA. When the policy is enabled, users must specify the exact commit SHA, for example, `/ok-to-test <sha>`, for approval. As a result, this ties the approval directly to a specific commit, preventing pipeline execution on any subsequent malicious force-pushed SHA.

Reviewer 2: @ochromy in place of "policy" I think it should be "config"

Author: Changed to "setting".

//+
//link:https://issues.redhat.com/browse/SRVKP-9136[SRVKP-9136]

.{tekton-hub}

Default database migration to PostgreSQL version 15::
The PostgreSQL database version used by the {tekton-hub} is migrated from version 13 to version 15 to address the upcoming end of life (EOL) for PostgreSQL 13. This upgrade ensures continued stability and support for the {tekton-hub}. Additionally, the process implements an automated transition from version 13 to version 15 for existing deployments.
//+
//link:https://issues.redhat.com/browse/SRVKP-9337[SRVKP-9337]

.{tekton-chains}

Flexible provenance and signing configuration::
With this update, you can choose to disable image signing while still enabling provenance generation and attestation signing. This enhancement helps provide more flexibility in managing security artifacts within your CI/CD pipelines.
+

> link:https://issues.redhat.com/browse/SRVKP-9137[SRVKP-9137]

Author: Yes, and I also removed it for the "New option to disable OCI image signing::". Accidentally left those in. Thank you for the catch.

A new `tkn` CLI command, `tkn pac cel`, allows administrators to interactively evaluate Common Expression Language (CEL) expressions against webhook payloads and headers. You can use the following syntax:
+
[source,terminal]
----
$ tkn pac cel -b <body.json> -H <headers.txt>
----
* `-b` or `--body`: Specify a path to a JSON body file. This is a webhook payload.
* `-H` or `--headers`: Specify a path to headers file. This can be a plain text, a JSON file, or a gosmee script.
* `-p` or `--provider`: Specify the provider. This can be `github`, `gitlab`,  `bitbucket-cloud`, `bitbucket-datacenter`, `gitea`, or `auto` for automatic detection from the payload.

+
Key capabilities include:
* Interactive mode: Provides a prompt in the terminal to type CEL expressions, with tab completion for variables and payload fields.
* Variable access using:
** direct variables, such as `event` or `target_branch`,
** webhook payload fields, for example, `body.action`,
** HTTP headers, for example, `headers['X-GitHub-Event']`,
** {pac} parameters, for example, `pac.revision`.
* Debugging: Quickly test and debug CEL expressions used in `PipelineRun` resource configurations against real webhook data.
+

> link:https://issues.redhat.com/browse/SRVKP-8026[SRVKP-8026]

Reviewer 2: @ochromy why we have added JIRA link in the few RN items?

Author: The new SSG requires including the links to the eng Jiras in Known and Fixed Issues. 
The link here should be commented out. Thank you again for the catch.

+
[source, yaml]
----
apiVersion: tekton.dev/v1
  kind: PipelineRun
  metadata:
    name: example-with-timeout-override
  spec:
    pipelineRef:
      name: my-pipeline
    timeouts:
      pipeline: "2h"
    taskRunSpecs:
      - pipelineTaskName: build-task
        timeout: "30m"
      - pipelineTaskName: test-task
        timeout: "1h"
----
+
This allows finer-grained control over task execution duration without affecting the overall `PipelineRun` timeout.

//https://issues.redhat.com/browse/SRVKP-8880
Resolver caching for bundle, Git, and cluster resolvers::
+
With this update, resolver caching is supported for bundle, Git, and cluster resolvers. This helps reduce redundant fetches, minimize external API calls, and improve pipeline execution reliability, especially when external services impose rate limits or are temporarily unavailable.
+
* You can configure caching using your `TektonConfig` custom resource (CR), where you can set the cache size and adjust the time to live (TTL) value without restarting controllers:
+
[source,yaml]
----
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonConfig
metadata:
  name: config
spec:
  pipeline:
    options:
      configMaps:

>         resolver-cache-config:

Reviewer: ```suggestion
        git-resolver-config:
```

>         resolver-cache-config:

Reviewer 6: Yes we can mention that  

The resolver-cache-config sets global cache size and TTL (shared LRU cache across all resolvers). The bundleresolver-config, git-resolver-config and cluster-resolver-config are set per-resolver default cache mode. Users can override at the TaskRun/PipelineRun level with the cache parameter as well.

>         resolver-cache-config:

Reviewer 2: @waveywaves @srivickynesh can you please and resolve this comment?

>         resolver-cache-config:

Reviewer: looks good

          data:
            max-size: "1000"
            ttl: "5m"
#...
----
+
** `max-size`: the maximum number of cached entries. The default value is `"1000"`.
** `ttl`: the time to live of the cache entry. The default value is `"5m"`.
+
* Caching mode can be set globally or on the task-level:
+
[source,yaml]
----
#...
data:
  cache: "auto"
  #...
----
+

> ** The available modes are `auto` (cache only immutable references), `always` (cache everything), and `never` (disable caching). The default value is `auto`.

Reviewer: except cluster resolver which uses `never` by default

Reviewer 2: @ochromy is this change incorporated?

          #...
----
+
Log snippets are automatically truncated to 65,000 characters, preventing failures when posting check-run updates to the GitHub API.

//https://issues.redhat.com/browse/SRVKP-8908
GitLab commit status set on the source and target project by default::
With this update, {pac} attempts to set the commit status on both the source and target `upstream` projects in GitLab. If both attempts fail due to permissions issues, the system falls back to posting a comment. This helps ensure that the commit status is reported in the most relevant location for merge requests.

//https://issues.redhat.com/browse/SRVKP-8902
Improved error message for GitLab private repository access failures::
With this update, {pac} proactively checks whether the configured GitLab token has the required read access to the source repository when processing merge requests from private forks. If the token lacks the necessary `read_repository` scope, {pac} fails early with a clear, actionable error message. This helps ensure that permission issues are easier to identify and prevents pipelines from failing unexpectedly later in the process.

//https://issues.redhat.com/browse/SRVKP-8577

> AI integration for automated pull request analysis::

Reviewer 3: @ochromy this is also "Technology Preview" feature

Cosign v2.6.0 update affects keyless signing::
With this update, {tekton-chains} uses Cosign version 2.6.0, which no longer accepts HS256 JWT tokens for keyless signing. If your private OIDC provider uses HS256 tokens for authentication, you must switch to RS256 before upgrading to this release. If you use perform key-based signing, or use a private OIDC provider already configured with RS256, you are not affected by this change.

[id="pipelines-known-issues-1-21-0"]
== Known issues

.User interface

Duplicate Pipelines navigation entry in the OpenShift Console::
During the transition from static to dynamic console plugins, the OpenShift Console might temporarily display two *Pipelines* entries in the navigation menu. This is a UI-only issue and does not affect pipeline execution or data.
+
To work around this problem, apply the corresponding updates to the OpenShift Console and the {pipelines-title} Operator. Users running older {rhosp} versions should coordinate upgrades with {OCP} to prevent a temporary disappearance of the Pipelines menu.
+
link:https://issues.redhat.com/browse/SRVKP-10006[SRVKP-10006]

.{pac}

Unclear error messages for the cel subcommand::

> When using the `opc pac cel` subcommand, providing no values for the `--body` or `--header` flags results in unclear error messages. This behavior makes it difficult to diagnose configuration issues when testing Common Expression Language (CEL) expressions.

Reviewer 3: this was find during unreleased build tests and the feature is being shipped in this very release (1.20) so should we include release note for a bug which is found during testing??

+
To work around this problem, ensure that both the `--body` and `--header` flags are correctly specified with valid file paths when running the subcommand.
+
link:https://issues.redhat.com/browse/SRVKP-9400[SRVKP-9400]

The cel subcommand panics with invalid GitLab data::

> The `cel` subcommand panics when invalid GitLab headers and body values are provided. This prevents the command from exiting gracefully during webhook evaluation.

Reviewer 3: same for this, it is found during release testing so should we include this in release notes?
cc: @srivickynesh @divyansh42

Reviewer 3: @ochromy please let me know when you're done with my changes

+
link:https://issues.redhat.com/browse/SRVKP-9200[SRVKP-9200]

GitLab API compatibility fix for canceled PipelineRuns::
Before this update, an issue in the GitLab provider system did not properly recognize canceled `PipelineRuns` due to a spelling mismatch between "cancelled" and "canceled", which is expected by the GitLab API. This issue caused GitLab merge requests to automatically merge even though the associated `PipelineRun` was canceled. With this update, this issue is fixed. An explicit mapping from "cancelled" to "canceled" for GitLab API compatibility helps ensure that cancelled `PipelineRuns` are correctly reported and merge requests remain open as expected.
+
link:https://issues.redhat.com/browse/SRVKP-9050[SRVKP-9050]

Placeholder variable evaluation no longer fails when data sources are missing::

> Before this update, placeholder variable evaluation failed when either the event payload or headers were missing. This failure occurred because the evaluation logic attempted to evaluate `body.\*`, `headers.\*`, and `files.\*` placeholders together. With this update, the evaluation logic processes these placeholders independently. As a result, each placeholder works if its corresponding data is present.

Reviewer 3: ```suggestion
Before this update, placeholder variable evaluation failed when either the event `payload` or `headers` were missing. This failure occurred because the evaluation logic attempted to evaluate `body.\*`, `headers.\*`, and `files.\*` placeholders together. With this update, the evaluation logic processes these placeholders independently. As a result, each placeholder works if its corresponding data is present.
```
